lib/money/money_parser.rb

```diff
@@ -1,56 +1,140 @@
+# frozen_string_literal: true
+
 # Parse an amount from a string
 class MoneyParser
-  MARKS = "\.,·’"
-  EXTRA_MARKS = "\s˙'"
+  class MoneyFormatError < ArgumentError; end
+
+  MARKS = %w[. , · ’ ˙ '] + [' ']
+
+  ESCAPED_MARKS = Regexp.escape(MARKS.join)
+  ESCAPED_NON_SPACE_MARKS = Regexp.escape((MARKS - [' ']).join)
+  ESCAPED_NON_DOT_MARKS = Regexp.escape((MARKS - ['.']).join)
+  ESCAPED_NON_COMMA_MARKS = Regexp.escape((MARKS - [',']).join)
+
+  NUMERIC_REGEX = /(
+    [\+\-]?
+    [\d#{ESCAPED_NON_SPACE_MARKS}][\d#{ESCAPED_MARKS}]*
+  )/ix
+
+  # 1,234,567.89
+  DOT_DECIMAL_REGEX = /\A
+    [\+\-]?
+    (?:
+      (?:\d+)
+      (?:[#{ESCAPED_NON_DOT_MARKS}]\d{3})+
+      (?:\.\d+)?
+    )
+  \z/ix
+
+  # 1.234.567,89
+  COMMA_DECIMAL_REGEX = /\A
+    [\+\-]?
+    (?:
+      (?:\d+)
+      (?:[#{ESCAPED_NON_COMMA_MARKS}]\d{3})+
+      (?:,\d+)?
+    )
+  \z/ix
 
-  def self.parse(input, currency = nil)
-    new.parse(input, currency)
+  # 12,34,567.89
+  INDIAN_NUMERIC_REGEX = /\A
+    [+\-]?
+    (?:
+      (?:\d+)
+      (?:\,\d{2})+
+      (?:\,\d{3})
+      (?:.\d+)?
+    )
+  \z/ix
+
+  # 1,1123,4567.89
+  CHINESE_NUMERIC_REGEX = /\A
+    [+\-]?
+    (?:
+      (?:\d+)
+      (?:\,\d{4})+
+      (?:.\d+)?
+    )
+  \z/ix
+
+  def self.parse(input, currency = nil, **options)
+    new.parse(input, currency, **options)
   end
 
-  def parse(input, currency = nil)
-    amount = extract_money(input.to_s, currency)
+  def parse(input, currency = nil, strict: false)
+    amount = extract_money(input.to_s, currency, strict)
     Money.new(amount, currency)
   end
 
   private
 
-  def extract_money(input, currency)
-    return '0' if input.empty?
+  def extract_money(input, currency, strict)
+    if input.empty?
+      return '0'
+    end
+
+    number = input.scan(NUMERIC_REGEX).flatten.first.try!(:strip)
+
+    unless number
+      raise MoneyFormatError, "invalid money string: #{input}" if strict
 
-    amount = input.scan(/(-?[\d#{MARKS}][\d#{MARKS}#{EXTRA_MARKS}]*)/).first
-    return '0' unless amount
-    amount = amount.first.tr(EXTRA_MARKS, '')
+      Money.deprecate("invalid money strings will raise in the next major release \"#{input}\"")
+      return '0'
+    end
 
-    *other_marks, last_mark = amount.scan(/[#{MARKS}]/)
-    return amount unless last_mark
+    marks = number.scan(/[#{ESCAPED_MARKS}]/).compact
+
+    if marks.empty?
+      return number
+    end
+
+    if marks.size == 1
+      return normalize_number(number, marks, currency)
+    end
+
+    if amount = number[DOT_DECIMAL_REGEX] || number[INDIAN_NUMERIC_REGEX] || number[CHINESE_NUMERIC_REGEX]
+      return amount.tr(ESCAPED_NON_DOT_MARKS, '')
+    end
+
+    if amount = number[COMMA_DECIMAL_REGEX]
+      return amount.tr(ESCAPED_NON_COMMA_MARKS, '').sub(',', '.')
+    end
+
+    raise MoneyFormatError, "invalid money string: #{input}" if strict
+
+    Money.deprecate("invalid money strings will raise in the next major release \"#{input}\"")
```

Comment:
Should we not show this deprecate message if strict is `true`?

Reply:
We only show this message if we don't raise the `MoneyFormatError`

Reply:
But it's already raising an exception if they're passing `strict` true. Isn't that already proving that they're ready to deal with the change in the next major release?

Reply:
Nevermind, forgot this method is returning early on successes, not on failures.